File: SMI-SDK/Java SDK/src/pt/sapo/gis/trace/appender/Appender.java
package pt.sapo.gis.trace.appender;

import pt.sapo.gis.exception.EventCallException;
import java.util.List;
import pt.sapo.gis.trace.TraceSeverityEventFilter;
import pt.sapo.gis.ICallback;
import pt.sapo.gis.trace.Entry;
import pt.sapo.gis.trace.EntrySeverityEventFilter;
import pt.sapo.gis.trace.Trace;
import pt.sapo.gis.trace.TraceManager;
import pt.sapo.gis.trace.configuration.Event;
import pt.sapo.gis.trace.configuration.EventType;
import pt.sapo.gis.trace.configuration.SeverityType;

/**
 * @author GIS Team
 * Base class to appenders
 */
public abstract class Appender {

    protected EntrySeverityEventFilter entrySeverityFilter = null;
    protected TraceSeverityEventFilter traceSeverityFilter = null;
    
    /**
     * Entry events callbacks
     */ 
    protected ICallback<Entry> entryCallback = new ICallback<Entry>() {

        @Override
        public void call(Entry e) throws EventCallException {
            entryEventHandler(e);
        }
    };
    
    /** 
     * Trace events callbacks
     */
    protected ICallback<Trace> traceCallback = new ICallback<Trace>() {

        @Override
        public void call(Trace t) throws EventCallException {
            traceEventHandler(t);
        }
    };

    /**     
     * Register the appender on trace events indicated on configuration object
     * @param config the appender configuration object
     */
    public Appender(pt.sapo.gis.trace.configuration.Appender config) {
        if (config.getEvents() != null) {
            for (Event event : config.getEvents().getList()) {
                List<SeverityType> severities = null;

                if (event.getFilters() != null) {
                    severities = event.getFilters().getSeverityFilters();
                }

                if (event.getType().equals(EventType.ENTRY)) {
                    if (severities != null) {
                        entrySeverityFilter = new EntrySeverityEventFilter(severities);
                    }
                    TraceManager.getOnEntryEvent().registerListener(entryCallback, entrySeverityFilter);
                }
                if (event.getType().equals(EventType.TRACE)) {
                    if (severities != null) {
                        traceSeverityFilter = new TraceSeverityEventFilter(severities);
                    }
                    TraceManager.getOnTraceEvent().registerListener(traceCallback, traceSeverityFilter);
                }
            }
        }
    }

    protected abstract void entryEventHandler(Entry e) throws EventCallException;

    protected abstract void traceEventHandler(Trace t) throws EventCallException;
}

File: SMI-SDK/.NET SDK/Library/gis/trace/appender/Appender.cs
namespace pt.sapo.gis.trace.appender
{
    using System.Linq;
    using System.Collections.Generic;
    using pt.sapo.gis.trace.configuration;
    using System;

    /**
     * @author GIS Team
     * Base class to appenders
     */
    public abstract class Appender : IDisposable
    {
        private GenericEventHandler<Entry> onEntry = null;
        private GenericEventHandler<Trace> onTrace = null;
        protected Dictionary<eventType, severityType[]> severities;

        public AppenderConfigProperties Properties { get; protected set; }

        /// <summary>
        /// Register the appender on trace events indicated on configuration object
        /// </summary>
        /// <param name="config">the appender configuration object</param>
        /// <exception cref="IntializatonException">When fail to initialize.</exception>
        public Appender(pt.sapo.gis.trace.configuration.appender config)
        {
            Properties = new AppenderConfigProperties(config.property);
            severities = config.events.Where(e => e.filters != null).ToDictionary(e => e.type, e => e.filters);
            foreach(@event evt in config.events){                
                switch(evt.type) {
                    case eventType.entry:
                        if (evt.filters == null || evt.filters.Length == 0)
                            onEntry = e => OnEntry(e);                            
                        else 
                            onEntry = e => { if(evt.filters.Contains(e.Severity)) OnEntry(e); };
                        TraceManager.OnEntry += onEntry;
                        break;
                    case eventType.trace:
                        if (evt.filters == null || evt.filters.Length == 0)
                            onTrace = t => OnTrace(t);
                        else 
                            onTrace = t => { if(t.GetSeverityTypes().Any(s => evt.filters.Contains(s))) OnTrace(t); };
                        TraceManager.OnTrace += onTrace;
                        break;
                }
            }
            
        }

        public abstract void OnEntry(Entry e);
        public abstract void OnTrace(Trace t);

        public void Dispose()
        {
            if (onEntry != null)
                TraceManager.OnEntry -= onEntry;
            if (onTrace != null)
                TraceManager.OnTrace -= onTrace;
        }

        ~Appender()
        {
            Dispose();
        }
    }
}
